### PR TITLE
Merge s9 soc support

### PR DESCRIPTION
**Pull request by jameshilliard**
Adds support for these miner models:
R4 S9 T9 T9P

tested on a Xilinx based S9

### CODE REVIEW

**Comment by ckolivas** [comment]
Nice, but unfortunately a totally screwy merge duplicating files just for bitmain convenience instead of using the existing files and properly merging the code.

**Comment by ckolivas** [comment]
Sha2 looks pretty pointless too.

File: driver-btm-soc.c [hunk 1]
```diff
@@ -0,0 +1,11793 @@
...
+        crc_data[1] = (unsigned char)((crc >> 0) & 0x00ff);
+
+        while(retry_count++<3)
+        {
+            //  pthread_mutex_lock(&iic_mutex);
+            T9_plus_write_pic_iic(false, false, 0x0, which_iic, PIC_COMMAND_1);
+            T9_plus_write_pic_iic(false, false, 0x0, which_iic, PIC_COMMAND_2);
+            T9_plus_write_pic_iic(false, false, 0x0, which_iic, length);
+            T9_plus_write_pic_iic(false, false, 0x0, which_iic, ERASE_PIC_APP_PROGRAM);
+            T9_plus_write_pic_iic(false, false, 0x0, which_iic,
```
Comment (jameshilliard) [inline driver-btm-soc.c]: looks like sha2-soc is required due to this.

Comment (jameshilliard) [inline driver-btm-soc.c]: Looks like sha2-soc is required due to this.

**Comment by jameshilliard** [comment]
Looks like sha2-soc is required due to [this](https://github.com/ckolivas/cgminer/blob/f749b22a45a9e558260e85c65f96428b8a4f3ee2/driver-btm-soc.c#L11034).

**Comment by ckolivas** [comment]
Yeah but sha2-soc still duplicates functionality for no reason. Only the custom functions should be added.

**Comment by bevinyxp** [comment]
How can I mining in asicboost mode? I connect slushpool and run the cgminer  , but can not find the stratum extension for asicboost. Is that anything I should do before running this software?

**Comment by ckolivas** [comment]
S9s don't have asicboost functionality that you can enable at the hardware level. Even if cgminer uses the asicboost work model it doesn't change the S9's efficiency.

**Comment by bevinyxp** [comment]
Hi ckolivas, thank you for your response. I am working on FPGA verification for asicboost not use S9.  I encountered a problem: how to make cgminer work on asicboost mode? could you please give me some advice? As I know, the common btc mining port(such as 3333) of slushpool support stratum extension, but I can not find some fields like vmak that use for asicboost when I connect this pool.

**Comment by jameshilliard** [comment]
> Yeah but sha2-soc still duplicates functionality for no reason. Only the custom functions should be added.

I removed sha2-soc and put the custom functions it uses in the driver.

> S9s don't have asicboost functionality that you can enable at the hardware level. Even if cgminer uses the asicboost work model it doesn't change the S9's efficiency.

This may not be accurate, AFAIK the ASIC chips should support it. It probably would require a custom FPGA firmware to properly enable though.

**Comment by sequencer** [comment]
I'm working on asicboost and FPGA.bin replacement.
maybe another one or two weeks.

**Pull request merged**